clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -538,51 +594,183 @@ private void updateGroupSubscription(Set<String> topics) {
+                    revokedPartitions);
 
-                // also clear the assigned partitions since all have been revoked
                 subscriptions.assignFromSubscribed(Collections.emptySet());
+            }
+        } else {
+            switch (protocol) {
+                case EAGER:
+                    // revoke all partitions
+                    revokedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                    exception = invokePartitionsRevoked(revokedPartitions);
 
-                break;
+                    subscriptions.assignFromSubscribed(Collections.emptySet());
 
-            case COOPERATIVE:
-                break;
+                    break;
+
+                case COOPERATIVE:
+                    // only revoke those partitions that are not in the subscription any more.
+                    Set<TopicPartition> ownedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+                    revokedPartitions = ownedPartitions.stream()
+                        .filter(tp -> !subscriptions.subscription().contains(tp.topic()))
+                        .collect(Collectors.toSet());
+
+                    if (!revokedPartitions.isEmpty()) {
+                        exception = invokePartitionsRevoked(revokedPartitions);
+
+                        ownedPartitions.removeAll(revokedPartitions);
+                        subscriptions.assignFromSubscribed(ownedPartitions);
+                    }
+
+                    break;
+            }
         }
 
+
         isLeader = false;
         subscriptions.resetGroupSubscription();
+
+        if (exception != null) {
+            throw new KafkaException("User rebalance callback throws an error", exception);
+        }
     }
 
+    private void resetAssignment(boolean lostPartitions) {
+        Set<TopicPartition> droppedPartitions = new HashSet<>(subscriptions.assignedPartitions());
+
+        if (subscriptions.partitionsAutoAssigned() && !droppedPartitions.isEmpty()) {
+
+            Exception e;
+
+            if (lostPartitions) {
+                e = invokePartitionsLost("reset generation", droppedPartitions);
+            } else {
+                e = invokePartitionsRevoked(droppedPartitions);
+            }
+
+            subscriptions.assignFromSubscribed(Collections.emptySet());
+
+            if (e != null) {
+                throw new KafkaException("User rebalance callback throws an error", e);
+            }
+        }
+    }
+
+    /**
+     * @throws KafkaException if the rebalance callback throws exception
+     */
+    @Override
+    synchronized void resetGenerationOnResponseError(ApiKeys api, Errors error) {
+        // heartbeat error handling is executed by heartbeat thread, in which case
+        // we would not drop partitions and trigger rebalance callback as it should
+        // only be triggered by the caller thread
+        if (api != ApiKeys.HEARTBEAT) {
```

Comment:
Hmm.. I am not sure this is sufficient. Any of the responses could return from the heartbeat thread.